packages/graph/hash_graph/lib/graph/src/store/postgres/ontology/entity_type/mod.rs

```diff
@@ -1,15 +1,120 @@
 mod resolve;
 
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    future::Future,
+    pin::Pin,
+};
+
 use async_trait::async_trait;
 use error_stack::{IntoReport, Result, ResultExt};
+use futures::{stream, FutureExt, StreamExt, TryStreamExt};
 use tokio_postgres::GenericClient;
-use type_system::EntityType;
+use type_system::{uri::VersionedUri, EntityType};
 
 use crate::{
-    ontology::{AccountId, PersistedOntologyIdentifier},
-    store::{AsClient, EntityTypeStore, InsertionError, PostgresStore, UpdateError},
+    ontology::{
+        AccountId, EntityTypeTree, PersistedDataType, PersistedEntityType, PersistedLinkType,
+        PersistedOntologyIdentifier, PersistedPropertyType,
+    },
+    store::{
+        crud::Read,
+        postgres::{context::PostgresContext, PersistedOntologyType},
+        query::Expression,
+        AsClient, EntityTypeStore, InsertionError, PostgresStore, QueryError, UpdateError,
+    },
 };
 
+impl<C: AsClient> PostgresStore<C> {
+    /// Internal method to read a [`PersistedEntityType`] into a [`HashMap`].
+    ///
+    /// This is used to recursively resolve a type, so the result can be reused.
+    #[expect(
+        clippy::too_many_arguments,
+        reason = "Output is created in place and needs to be reused later"
```

Comment:
We could probably unify the references with their depth in a tuple (HashMap, u8), and that'd get rid of the lint and tie the things together. I'm not _sure_ it would be nicer though.

Reply:
I tried that but it's trickier than expected. I experimented with 
```rust
pub struct ReferenceResolverContext<'r, V, T> {
    pub references: &'r mut HashMap<V, T>,
    pub query_depth: u8,
}
```

We require the `references` to be a unique reference, but the query depth shouldn't change, so the reference has to be captured inside of the struct. This however implies, that we cannot pass `&mut ReferenceResolverContext` (depth shouldn't change) and we cannot pass `&ReferenceResolverContext`, as then the unique reference is behind a shared reference. Capturing the unique reference means, that the type cannot implement `Clone`, so we need to reconstruct it each time we want to pass it further inside of a for loop (the for-loop permits us passing it by value).

I definitely hear you, but the current approach is more readable 😢 
We could do the same as we do in TS and pass a struct everywhere, but I don't know if this helps, especially as Rust does not have anonymous structs.